lale/lib/lale/aggregate.py

```diff
@@ -12,16 +12,81 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import ast
+
+import pandas as pd
+
 import lale.docstrings
 import lale.operators
 
+try:
+    import pyspark
+    from pyspark.sql.dataframe import DataFrame as spark_df
+
+    spark_installed = True
+
+except ImportError:
+    spark_installed = False
+
+
+def _is_pandas_df(df):
+    return isinstance(df, pd.core.groupby.generic.DataFrameGroupBy) or isinstance(
+        df, pd.DataFrame
+    )
+
+
+def _is_spark_df(df):
+    return isinstance(df, pyspark.sql.group.GroupedData) or isinstance(df, spark_df)  # type: ignore
+
+
+def _is_ast_subscript(expr):
+    return isinstance(expr, ast.Subscript)
+
+
+def _is_ast_attribute(expr):
+    return isinstance(expr, ast.Attribute)
+
 
 class _AggregateImpl:
-    def __init__(self, columns, group_by=[]):
-        self._hyperparams = {"columns": columns, "group_by": group_by}
+    def __init__(self, columns):
+        self.columns = columns
 
     def transform(self, X):
-        raise NotImplementedError()
+        if _is_pandas_df(X):
+            agg_col = []
+            agg_func = []
+            rename_col = []
+            agg_expr = {}
+
+            for new_col_name, expr in self.columns.items():
+                rename_col.append(new_col_name)
+                agg_func.append(expr._expr.func.id)
+                expr_to_parse = expr._expr.args[0]
+                if _is_ast_subscript(expr_to_parse):
+                    agg_col.append(expr_to_parse.slice.value.s)  # type: ignore
+                elif _is_ast_attribute(expr_to_parse):
+                    agg_col.append(expr_to_parse.attr)
+                else:
+                    raise ValueError(
+                        "Aggregate 'columns' parameter only supports subscript or dot notation for the key columns. For example, it.col_name or it['col_name']."
+                    )
+
+            zipped_lists = zip(agg_col, agg_func, rename_col)
+            sorted_pairs = sorted(zipped_lists)
+            tuples = zip(*sorted_pairs)
+            agg_col, agg_func, rename_col = [list(tuple) for tuple in tuples]
```

Comment:
this seems a bit roundabout.  Maybe instead, create a dictionary from
rename_col to tuples of the aggregate function and the aggregate column
(type: Dict[str, Tuple[str, str]])
while createing this, you can check before adding a new element to make sure that the set of rename columns is unique.
Creating this structure can be shared for both the pandas and (eventual) spark codepaths -- they both then iterated

Reply:
Thanks Avi, will surely work on this. Actually the PR is still WIP, my bad I did not mention it in the comments.